sympy/matrices/expressions/matpow.py

```diff
@@ -92,7 +91,12 @@ def _eval_transpose(self):
         return MatPow(base.T, exp)
 
     def _eval_derivative(self, x):
-        return Pow._eval_derivative(self, x)
+        assert not hasattr(x, "shape")
+        dbase = self.base.diff(x)
+        dexp = self.exp.diff(x)
+        if dexp != 0:
+            raise NotImplementedError("cannot derive matrix power if exponent depends on the derivation variable")
+        return self * (dbase * self.exp/self.base)
```

Comment:
hope I got this correct!

Reply:
This is not correct because it divides by `self.base` which is a matrix:
```python
In [16]: M = MatPow(Matrix([[x, 0], [0, x]]), 2)

In [17]: M
Out[17]: 
      2
⎡x  0⎤ 
⎢    ⎥ 
⎣0  x⎦ 

In [18]: M.diff(x)
...
sympy/sympy/matrices/expressions/matexpr.py:163, in MatrixExpr.__rtruediv__(self, other)
    160 @_sympifyit('other', NotImplemented)
    161 @call_highest_priority('__truediv__')
    162 def __rtruediv__(self, other):
--> 163     raise NotImplementedError()
```
Suppose we have the simplest case `d/dx (M(x)**2)`:
```
(M(x)^2)' = (M(x)*M(x))' = M'(x)*M(x) + M(x)*M'(x)
```
If `M` and `M'` commute then we can simplify this to:
```
(M(x)^2)' = 2*M(x)*M'(x)
```
which is the chain/product rule in the commutative case.

The formula used here says that:
```
(M(x)^2)' = M(x) * (M'(x) * 2/M(x))
```
Taking out the division by a matrix the formula would be correct for the commutative case but we can't assume that `M` and `M'` commute:
```python
In [26]: M = Matrix([[0, x], [1, 0]])

In [27]: M
Out[27]: 
⎡0  x⎤
⎢    ⎥
⎣1  0⎦

In [28]: M*M.diff(x)
Out[28]: 
⎡0  0⎤
⎢    ⎥
⎣0  1⎦

In [29]: M.diff(x)*M
Out[29]: 
⎡1  0⎤
⎢    ⎥
⎣0  0⎦
```
The general non-commutative product rule holds:
```python
In [30]: M**2
Out[30]: 
⎡x  0⎤
⎢    ⎥
⎣0  x⎦

In [31]: (M**2).diff(x)
Out[31]: 
⎡1  0⎤
⎢    ⎥
⎣0  1⎦

In [32]: M*M.diff(x) + M.diff(x)*M
Out[32]: 
⎡1  0⎤
⎢    ⎥
⎣0  1⎦
```
That could be used when the exponent is an explicit integer.

If the matrix is invertible then we can in principle its logarithm can be used:
```
M(x)**f(x) = exp(log(M(x)))**f(x)
```
I think if `f(x)` is real then this is
```
(M(x)**f(x))' = exp(f(x)*log(M(x)))' = (f(x)*log(M(x))' * exp(f(x)*log(M(x)))
   = (f'(x)*log(M(x)) + f(x)*log(M(x))') * M(x)**f(x)
```
I'm not totally sure about `log(M(x))'`:
```
A(x) = log(M(x))'  -->  M(x) = exp(A(x))
M'(x) = A'(x) * exp(A(x))
A'(x) = exp(A(x))**-1 * M'(x)
log(M(x))' = M(x)**-1 * M'(x)
```
Then
```
(M(x)**f(x))' = (f'(x)*log(M(x)) + f(x)*M(x)**-1 * M'(x)) * M(x)**f(x)
```
If `f'(x)` is zero then this is:
```
(M(x)**f(x))' = f(x) * M(x)**-1 * M'(x) * M(x)**f(x)
```
This is similar to the formula used here but has `M(x)**-1` rather than `1/M(x)` and also has the factors in a different order. I think that I have assumed only that `M(x)` is invertible and `f(x)` is real and constant but not that `M` and `M'` commute...

Reply:
> If `f'(x)` is zero then this is:
> ```
> (M(x)**f(x))' = f(x) * M(x)**-1 * M'(x) * M(x)**f(x)
> ```

This formula does not hold:
```
In [21]: M = Matrix([[0, x], [1, 0]])

In [22]: (M**3).diff(x)
Out[22]: 
⎡0  2⋅x⎤
⎢      ⎥
⎣1   0 ⎦

In [23]: 3*M**-1 * M.diff(x) * M**3
Out[23]: 
⎡0  0⎤
⎢    ⎥
⎣3  0⎦
```
I think I must have assumed that `M` and `M'` commute somewhere. If they do commute then it holds I think:
```python
In [27]: M = Matrix([[0, x], [x, 0]])

In [28]: (M**3).diff(x)
Out[28]: 
⎡         2⎤
⎢ 0    3⋅x ⎥
⎢          ⎥
⎢   2      ⎥
⎣3⋅x    0  ⎦

In [29]: 3*M**-1 * M.diff(x) * M**3
Out[29]: 
⎡         2⎤
⎢ 0    3⋅x ⎥
⎢          ⎥
⎢   2      ⎥
⎣3⋅x    0  ⎦
```